staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/established_controller.go

```diff
@@ -0,0 +1,101 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package apiserver
+
+import (
+	"k8s.io/apimachinery/pkg/util/wait"
+	"time"
+	"k8s.io/apiextensions-apiserver/pkg/apis/apiextensions"
+	"k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/kubernetes/staging/src/k8s.io/client-go/util/workqueue"
+	"k8s.io/apiextensions-apiserver/pkg/client/clientset/internalclientset"
+	"github.com/golang/glog"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+)
+
+// EstablishingController controls how and when CRD is established.
+type EstablishingController struct {
+	client *internalclientset.Clientset
+	queue workqueue.DelayingInterface
+}
+
+// NewEstablishingController creates new EstablishingController.
+func NewEstablishingController(client *internalclientset.Clientset) *EstablishingController {
+	return &EstablishingController{
+		client: client,
+		queue: workqueue.NewDelayingQueue(),
+	}
+}
+
+func (ec *EstablishingController) Run(stopCh <-chan struct{}) {
+	defer utilruntime.HandleCrash()
+	defer ec.queue.ShutDown()
+
+	glog.Infof("Starting NamingConditionController")
+	defer glog.Infof("Shutting down NamingConditionController")
+
+	// only start one worker thread since its a slow moving API and the naming conflict resolution bits aren't thread-safe
+	go wait.Until(ec.runWorker, time.Second, stopCh)
+
+	<-stopCh
+}
+
+func (ec *EstablishingController) runWorker() {
+	for ec.establishCRD() {
+	}
+}
+
+// EstablishCRD turns CRD into established and returns
+// is it successful or error
+func (ec *EstablishingController) establishCRD() bool {
+	// We will get back to this for{} :)
+		key, quit := ec.queue.Get()
+		if quit {
+			return false
+		}
+		defer ec.queue.Done(key)
+		crd, _ := ec.client.Apiextensions().CustomResourceDefinitions().Get(key.(string), v1.GetOptions{})
+
+		// Poll because if multiple servers try to change CRD (e.g. in HA) this can fail.
+		// That failure is normal, and so, we're trying to set it to Established again.
+		err := wait.Poll(500 * time.Millisecond, 3 * time.Second, func() (bool, error) {
```

Comment:
what do other controllers do in the error case? re-queue the key?

Reply:
In case of naming controller, yes.